src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,146 @@
+    private T[] elements;
+    private int size;
+
+    public ArrayList(int initialCapacity) {
+        if (initialCapacity > 0) {
+            elements = (T[]) new Object[initialCapacity];
+        } else {
+            throw new IllegalArgumentException("Illegal initial capacity: " + initialCapacity);
+        }
+    }
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
+        growIfNeed();
 
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Incorrect index " + index);
+        }
 
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            growIfNeed();
+            T[] newElements = (T[]) new Object[elements.length];
+            newElements[index] = value;
+            System.arraycopy(elements, 0, newElements, 1, size);
+            elements = newElements;
+            size++;
+        } else {
+            growIfNeed();
+            T[] newElements = (T[]) new Object[elements.length];
+            System.arraycopy(elements, 0, newElements, 0, index);
+            newElements[index] = value;
+            System.arraycopy(elements, index, newElements, index + 1, size - index + 1);
+            elements = newElements;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            growIfNeed();
+            elements[size] = list.get(i);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Incorrect index " + index);
+        }
+
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Incorrect index " + index);
+        }
 
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Incorrect index " + index);
+        }
+
+        T oldValue = elements[index];
+        if (index == size - 1) {
+            elements[index] = null;
+        } else if (index == 0) {
+            T[] newElements = (T[]) new Object[elements.length];
+            System.arraycopy(elements, 1, newElements, 0, size);
+            elements = newElements;
+        } else {
+            T[] newElements = (T[]) new Object[elements.length];
+            System.arraycopy(elements, 0, newElements, 0, index);
+            System.arraycopy(elements, index + 1, newElements, index, size - index);
```

Comment:
arraycopy is very heavy method, use it only once
you may perform copy having same array as source and destination